apps/root/src/pages/token-profile/historical-prices/index.tsx

```diff
@@ -0,0 +1,356 @@
+                {formatCurrencyAmount({ amount: value.amount.amount, token: value.token, intl })}
+              </Typography>
+              <Typography variant="bodySmallRegular" color={({ palette: { mode } }) => colors[mode].typography.typo3}>
+                (${formatUsdAmount({ intl, amount: value.amount.amountInUSD })})
+              </Typography>
+            </ContainerBox>
+          </ContainerBox>
+        ))
+      )}
+    </StyledTooltipContainer>
+  );
+};
+
+const PeriodMap: Record<Exclude<GraphContainerPeriods, GraphContainerPeriods.all>, TimeString> = {
+  [GraphContainerPeriods.day]: '30m',
+  [GraphContainerPeriods.week]: '3h',
+  [GraphContainerPeriods.month]: '12h',
+  [GraphContainerPeriods.year]: '1w',
+};
+
+const DATA_POINTS = 200;
+
+const TokenHistoricalPrices = ({ token }: TokenHistoricalPricesProps) => {
+  const mode = useThemeMode();
+  const priceService = usePriceService();
+  const providerService = useProviderService();
+  const [selectedPeriod, setSelectedPeriod] = React.useState<GraphContainerPeriods>(GraphContainerPeriods.month);
+  const prevPeriod = usePrevious(selectedPeriod);
+  const [isLoadingHistorical, setIsLoadingHistorical] = React.useState(true);
+  const [graphPrices, setGraphPrices] = React.useState<Partial<Record<GraphContainerPeriods, PriceResult[]>>>({});
+
+  const handlePeriodUpdate = (newPeriod: GraphContainerPeriods) => {
+    if (!graphPrices[newPeriod]) {
+      // When period changes, the useMemo renders no data before the isLoading is set to true by the useEffect
+      // By setting it here, we avoid the flickering of the graph
+      setIsLoadingHistorical(true);
+    }
+
+    setSelectedPeriod(newPeriod);
+  };
+
+  React.useEffect(() => {
+    const handleAllTimeframe = async () => {
+      const firstTimestamp = await providerService.getBlockTimestamp(token.chainId, 1n);
+      const currentTimestamp = Math.floor(Date.now() / 1000);
+
+      const totalDataPoints = Math.floor((currentTimestamp - firstTimestamp) / (24 * 60 * 60));
+
+      if (totalDataPoints < 30) {
+        // Fresh chain
+        return {
+          period: PeriodMap[GraphContainerPeriods.day],
+          span: DATA_POINTS,
+        };
+      }
+
+      let span = totalDataPoints;
+
+      let periodDays = 1;
+      while (span > DATA_POINTS) {
+        span = Math.floor(span / 3);
+        periodDays = periodDays * 3;
+      }
+
+      return {
+        period: `${periodDays}d` as TimeString,
+        span,
+      };
+    };
+
+    const fetchGraphPrices = async () => {
+      const priceResult = await priceService.getPricesForTokenGraph({
+        token,
+        ...(selectedPeriod === GraphContainerPeriods.all
+          ? await handleAllTimeframe()
+          : { period: PeriodMap[selectedPeriod], span: DATA_POINTS }),
+      });
+
+      setGraphPrices((prev) => ({ ...prev, [selectedPeriod]: priceResult }));
```

Comment:
see, here we dont need to have results per period, we could just have results and extend those when a period gets selected